src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -3016,6 +3033,122 @@ - (BOOL)_attributeAffectsDeviceConfiguration:(MTRAttributePath *)attributePath
     return NO;
 }
 
+- (void)_removeClusters:(NSSet *)clusterPathsToRemove
+{
+    os_unfair_lock_assert_owner(&self->_lock);
+
+    [_persistedClusters minusSet:clusterPathsToRemove];
+
+    for (MTRClusterPath * path in clusterPathsToRemove) {
+        [_persistedClusterData removeObjectForKey:path];
+        [_persistedClusters removeObject:path];
+        [_clusterDataToPersist removeObjectForKey:path];
+        [self.deviceController.controllerDataStore clearStoredClusterDataForNodeID:self.nodeID endpointID:path.endpoint clusterID:path.cluster];
+    }
+}
+
+- (void)_removeAttributes:(NSSet *)toBeRemovedAttributes fromCluster:(MTRClusterPath *)clusterPathToRemoveAttributesFrom
+{
+    if (toBeRemovedAttributes == nil || clusterPathToRemoveAttributesFrom == nil) {
+        return;
+    }
+    os_unfair_lock_assert_owner(&self->_lock);
+
+    for (NSNumber * attribute in toBeRemovedAttributes) {
+        [self _removeCachedAttribute:attribute fromCluster:clusterPathToRemoveAttributesFrom];
+    }
+    [_persistedClusterData removeObjectForKey:clusterPathToRemoveAttributesFrom];
+    [self.deviceController.controllerDataStore clearStoredClusterDataForNodeID:self.nodeID endpointID:clusterPathToRemoveAttributesFrom.endpoint clusterID:clusterPathToRemoveAttributesFrom.cluster];
```

Comment:
Tests are passing because _clusterDataToPersist has the right cluster data and is written back before the test asserts. Thats why they are passing. this is incorrect. Thanks for catching this.

Reply:
Removed. was a bug.